README.md

```diff
@@ -1 +1,252 @@
+   development mode.
+
+<table class="table table-condensed table-bordered table-params">
+  <colgroup>
+    <col class="col-param" />
+    <col class="param-description" />
+  </colgroup>
+  <thead>
+    <tr>
+      <th colspan="2">Attributes</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><code>name</code></td>
+      <td>
+        <code>Name, required</code>
+        <p>A unique name for this rule.</p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>srcs</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          List of .java source files that will be compiled and passed on the
+          classpath to the GWT compiler.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>resources</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          List of resource files that will be passed on the classpath to the GWT
+          compiler, e.g. <code>.gwt.xml</code>, <code>.ui.xml</code>, and
+          <code>.css</code> files.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>modules</code></td>
+      <td>
+        <code>List of strings, required</code>
+        <p>
+          List of fully-qualified names of modules that will be passed to the GWT
+          compiler. Usually contains a single module name corresponding to the
+          application's <code>.gwt.xml</code> file.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>pubs</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          Files that will be copied directly to the output war, such as static
+          HTML or image resources. Not interpreted by the GWT compiler.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>deps</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          List of other java_libraries on which the application depends. Both the
+          class jars and the source jars corresponding to each library as well as
+          their transitive dependencies will be passed to the GWT compiler's
+          classpath. These libraries may contain other <code>.gwt.xml</code>,
+          <code>.ui.xml</code>, etc. files as <code>resources</code>.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>output_root</code></td>
+      <td>
+        <code>String, optional</code>
+        <p>
+          Directory in the output war in which all outputs will be placed. By
+          default outputs are placed at the root of the war file.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>java_root</code></td>
+      <td>
+        <code>String, optional</code>
+        <p>
+          Directory relative to the workspace root that forms the root of the
+          Java package hierarchy (e.g. it contains a <code>com</code> directory).
+          By default this is `src/main/java`. If your Java files aren't under
```

Comment:
You can detect a common root for the input files. But it's not guaranteed to be the java root, since you can't read the `package` declaration in the file from the rules.

I'd recommend a few things here:

1. Allow more than one java root. Mandatory considering external repositories, e.g. `external/foo/src/java/main/`.
2. Make the default include `java` and `javatests` since we're exporting Google engineering culture 😉 
3. Add a check to make sure that files in the transitive closure are beneath the roots, using the following function to calculate the paths against which they're judged:

```python
def runpath(f):
  """Figures out the proper runfiles path for a file."""
  if f.path.startswith('bazel-out/') or f.path.startswith('bazel-genfiles/'):
    return f.short_path
  else:
    return f.path
```

Also looking at the design of these rules, I'm not 100% sure if they would work for large GWT applications, where you have GWT code spread out across multiple rules, in multiple repositories, some external, some local.